### PR TITLE
Allow to filter net devices by Infiniband Partition Key

### PR DESCRIPTION
**Pull request by almaslennikov**
This PR introduced a new kind of device selector based on [Infiniband Partition Key](https://docs.nvidia.com/networking/display/winof2v310lts/infiniband+network#src-106007664_InfiniBandNetwork-PKeys).
The PKey retrieval functionality is taken from sriovnet. This PR will be finalized with updated sriovnet dependency when the ([related PR there](https://github.com/k8snetworkplumbingwg/sriovnet/pull/69)) is merged

### CODE REVIEW

**Review by Eoghan1232 — APPROVED** [review]
@almaslennikov Thanks for the contribution!
this looks good to me, could you fix the lint issue as well? just remove the commented out line.

File: pkg/resources/pkey_selector.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+func NewPKeySelector(pKeys []string) types.DeviceSelector {
+	return &pkeySelector{pKeys: pKeys}
+}
+
+func (ds *pkeySelector) Filter(inDevices []types.HostDevice) []types.HostDevice {
+	filteredList := make([]types.HostDevice, 0)
+
+	for _, dev := range inDevices {
+		pKey := dev.(types.PciNetDevice).GetPKey()
+		if pKey != "" && contains(ds.pKeys, pKey) {
```
Comment (Eoghan1232) [inline pkg/resources/pkey_selector.go:21]: could we keep the names consistent? 
pKeys in some places, pkey in others, and PKeys in another.

**Comment by almaslennikov** [comment]
The new sriovnet dependency requires upgrading to go1.21. I've created a separate PR to update the toolchain: https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/527

File: pkg/resources/pkey_selector.go (hunk 1 above)
Comment (almaslennikov) [reply pkg/resources/pkey_selector.go]: Thanks for noticing! I've changed lowercase pkey to the snake case. pKey/PKey are both snake case and determined by the context

File: pkg/netdevice/pciNetDevice.go [hunk 2]
```diff
@@ -116,3 +116,13 @@ func (nd *pciNetDevice) GetDDPProfiles() string {
 func (nd *pciNetDevice) GetVdpaDevice() types.VdpaDevice {
 	return nd.vdpaDev
 }
+
+func (nd *pciNetDevice) GetPKey() string {
```
[the remark this answers is not in the document]
Comment (almaslennikov) [reply pkg/netdevice/pciNetDevice.go:131]: Moved PKey retrieval logic to pciNetDevice constructor

**Comment by adrianchiris** [comment]
@almaslennikov can you sort out lint test ?

**Review by Eoghan1232 — APPROVED** [review]
lgtm! thanks for fixing linting issue

File: pkg/resources/pKeySelector_test.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types/mocks"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("PKeySelector", func() {
+	Describe("PKey selector", func() {
+		Context("filtering", func() {
+			It("should return devices matching DDP profiles", func() {
```
Comment (adrianchiris) [inline pkg/resources/pKeySelector_test.go:15]: > should return devices matching DDP profiles

 fix the test msg :)

File: pkg/utils/sriovnet_provider.go [hunk 4]
```diff
@@ -52,3 +53,7 @@ func (defaultSriovnetProvider) GetNetDevicesFromAux(auxDev string) ([]string, er
 func (defaultSriovnetProvider) GetAuxNetDevicesFromPci(pciAddr string) ([]string, error) {
 	return sriovnet.GetAuxNetDevicesFromPci(pciAddr)
 }
+
+func (defaultSriovnetProvider) GetPKey(pciAddr string) (string, error) {
```
Comment (adrianchiris) [inline pkg/utils/sriovnet_provider.go:57]: can you name it after the sriovnet function ?

File: README.md [hunk 5]
```diff
@@ -337,6 +337,7 @@ These selectors are applicable when "deviceType" is "netDevice" (note: this is d
 | "rootDevices"  | N        | functions from PF matches list of PF PCI addresses                       | `string` list Default: `null`                       | "rootDevices": ["0000:86:00.0"] (See follow-up sections for some advance usage of "rootDevices") |
 | "linkTypes"    | N        | The link type of the net device associated with the PCI device           | `string` list Default: `null`                       | "linkTypes": ["ether"]                                                                           |
 | "ddpProfiles"  | N        | A map of device selectors                                                | `string` list Default: `null`                       | "ddpProfiles": ["GTPv1-C/U IPv4/IPv6 payload"]                                                   |
+| "pKeys"        | N        | Infiniband Partition Keys. Compatible only with linkTypes = infiniband   | `string` list Default: `null`                       | "pKeys": ["0x1", "0xABCD", "0x50"]                                                               |
```
Comment (adrianchiris) [inline README.md:340]: please add some info that it will match on the device's default pkey

**Review by adrianchiris — COMMENTED** [review]
overall this looks good. added a few small comments.

**Comment by almaslennikov** [comment]
@adrianchiris Thank you for the review! Addressed your comments

File: README.md (hunk 5 above)
Comment (adrianchiris) [reply README.md:340]: i dont see the update.

**Review by adrianchiris — APPROVED** [review]
LGTM ! . Thx @almaslennikov  :)

**Comment by adrianchiris** [comment]
@Eoghan1232 could you take a quick look at the lastest iteration ? if all is ok id like to merge this one

**Review by Eoghan1232 — APPROVED** [review]
latest iteration lgtm.

**Comment by Eoghan1232** [comment]
@adrianchiris are we waiting for all CI to pass or can we merge this?

**Comment by adrianchiris** [comment]
we can merge. we have issues with CI that need to be sorted internally.

**Pull request merged**